main/src/library/DecodingReader.flix

```diff
@@ -0,0 +1,123 @@
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+import java.nio.charset.CharsetDecoder
+import java.nio.ByteBuffer
+import java.nio.CharBuffer
+
+///
+/// Adapts a `Readable` of bytes to a `Readable` of chars.
+///
+/// `byteReader`: the underlying byte reader, an instance of `Readable where Readable.Elm[t] ~ Int8`.
+/// `decoder`: used to decode bytes read from `byteReader`.
+/// `leftovers`: stores bytes that have been consumed from `byteReader` but have not yet been decoded;
+///              either because they are lead bytes whose continuation bytes have not been read, or
+///              because `DecodingReader` overestimated the number of bytes needed to fill a char buffer.
+///
+struct DecodingReader[t, r] {
+    rc: Region[r],
+    byteReader: t,
+    decoder: CharsetDecoder,
+    mut leftovers: Array[Int8, r]
+}
+
+instance Readable[DecodingReader[t, rr]] with Readable[t] where Readable.Elm[t] ~ Int8 {
+
+    type Elm = Char
+
+    type Aef = Readable.Aef[t] + IO + rr
+
+    pub def read(buffer: Array[Char, rb], reader: DecodingReader[t, rr]): Result[IoError, Int32] \ rb + rr + Readable.Aef[t] + IO = DecodingReader.read(buffer, reader)
+
+}
+
+mod DecodingReader {
+    import java.nio.charset.CharsetDecoder
+    import java.nio.charset.CodingErrorAction
+    import java.nio.ByteBuffer
+    import java.nio.CharBuffer
+
+    ///
+    /// Wraps `byteReader` in a `DecodingReader`.
+    ///
+    pub def wrap(rc: Region[rr], charset: CharacterSet, byteReader: t): DecodingReader[t, rr] \ rr + IO =
+        let CharacterSet.CharacterSet(cs) = charset;
+        new DecodingReader @ rc {
+            rc = rc,
+            decoder = cs.newDecoder().onMalformedInput(CodingErrorAction.REPLACE),
+            byteReader = byteReader,
+            leftovers = Array.empty(rc, 0)
+        }
+
+    ///
+    /// Reads bytes from the underlying byteReader and decodes the result into `dest`.
+    ///
+    /// Returns `Ok(k)` to signify that `k` chars were successfully read and written to `dest`.
+    ///
+    /// Guarantees that `0 <= k <= length(dest)`.
+    ///
+    /// Returns `Err(e)` if some underlying error occurs.
+    ///
+    pub def read(dest: Array[Char, rb], reader: DecodingReader[t, rr]): Result[IoError, Int32] \ IO + rr + rb + Readable.Aef[t] with Readable[t] where Readable.Elm[t] ~ Int8 =
+        forM(
+            (src, endOfInput) <- {
+                // At most 4 bytes per char, guarantees we can fill dest if there are enough bytes to do so
+                let totalBytesNeeded = 4 * Array.length(dest);
+                readBytes(totalBytesNeeded, reader)
+            };
+            charsRead <- {
+                let charsRead = decode(src, dest, endOfInput, reader);
+
+                // charsRead == 0 would imply end of input to the caller.
+                // block again until end of input or we have at least one char to return
```

Comment:
This is not a busy wait/block because the underlying reader is not a busy waiter, right?

Reply:
It can be, ex reading from an inputstream on a tcp connection will block until there's some data. This is probably a sign of needing a better return type; since we have to do extra reading in order to comply with the "0 means EOF" contract.